components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -182,12 +195,33 @@ abstract class AbstractFetchDownloadService : Service() {
             startDownloadJob(download.id)
         }
 
+        notificationUpdateScope.launch {
+            while (isActive) {
+                delay(PROGRESS_UPDATE_INTERVAL)
+                updateDownloadNotificationProgress()
+            }
+        }
+
         return super.onStartCommand(intent, flags, startId)
     }
 
+    /***
+     * Android rate limits notifications being sent, so we must send them on a delay so that
+     * notifications are not dropped
+     */
+    @Synchronized
+    private fun updateDownloadNotificationProgress() {
+        for (download in downloadJobs.values) {
+            if (download.status != DownloadJobStatus.ACTIVE) { continue }
+            // We must be synchronized here to avoid the status getting set to PAUSED on this line
+            // and then overwriting that with an ongoing notification anyway
+            displayOngoingDownloadNotification(download.state, download.currentBytesCopied)
```

Comment:
I'm having trouble with this race condition. Do you have any suggestions about how to solve it? I thought putting a synchronized on this and all other state setting would be sufficient but it doesn't seem so 😕 

@pocmo

Reply:
Oh, sorry, this has fallen through the cracks. I can try to help with that this week. What is happening in that race condition? Notifications do not update or with wrong state?

Reply:
What’s happening is a user presses PAUSE, the notification then gets a timed update to display the current progress (ACTIVE state), so the “pause” action gets eaten but the download is still paused and stuck in a state where the user sees it as “frozen”